src/masternodeman.cpp

```diff
@@ -644,6 +671,401 @@ void CMasternodeMan::ProcessMessage(CNode* pfrom, std::string& strCommand, CData
         }
         // smth weird happen - someone asked us for vin we have no idea about?
         LogPrint("masternode", "CMasternodeMan::ProcessMessage -- DSEG -- No invs sent to peer %d\n", pfrom->id);
+
+    } else if (strCommand == NetMsgType::MNVERIFY) { // Masternode Verify
+
+        CMasternodeVerification mnv;
+        vRecv >> mnv;
+
+        if(mnv.vchSig1.empty()) {
+            // CASE 1: someone asked me to verify myself /IP we are using/
+            SendVerifyReply(pfrom, mnv);
+        } else if (mnv.vchSig2.empty()) {
+            // CASE 2: we _probably_ got verification we requested from some masternode
+            ProcessVerifyReply(pfrom, mnv);
+        } else {
+            // CASE 3: we _probably_ got verification broadcast signed by some masternode which verified another one
+            ProcessVerifyBroadcast(pfrom, mnv);
+        }
+    }
+}
+
+void CMasternodeMan::DoFullVerificationStep()
+{
+    if(activeMasternode.vin == CTxIn()) return;
+
+    int nCountInitial = -1;
+    int nCountMax = std::max(10, (int)vMasternodes.size() / 100); // verify at least 10 masternode at once but at most 1% of all known masternodes
+    int nCount = nCountInitial;
+
+    std::vector<std::pair<int, CMasternode> > vecMasternodeRanks = GetMasternodeRanks(pCurrentBlockIndex->nHeight - 1, MIN_POSE_PROTO_VERSION);
+    std::vector<CMasternode*> vSortedByAddr;
+
+    LOCK(cs);
+
+    BOOST_FOREACH(CMasternode& mn, vMasternodes) {
+        vSortedByAddr.push_back(&mn);
+    }
+
+    // send verify requests to up to nCountMax masternodes starting next to self (by rank)
+    // but only if we are in top MAX_POSE_RANK
```

Comment:
Agree about "unfair" resource usage, however that's only the case in one single block. On avg they should even imo but that's not trivial to proof so I think it's better to fix it :) However `(MAX_POSE_RANK * own rank)` can easily go out of bounds on testnet so I think `(MAX_POSE_RANK + own rank)` would be a better option here.